### PR TITLE
#24146 Fix Integration test runs and crossover between tests

### LINKED ISSUE #24146: Fix Integration test runs and crossover between tests
This pull request combines quite a few fixes and enhancements to help to diagnose problems and interactions between integration tests.   It should result in a consistent passing set of Integration tests and tools to help solve similar problems in the future.

Tools

Introduction of Awaitability library to replace Thread.sleep for tests that rely on asynchronous behavior and need to poll.
3 Checkers that can be turned on to validate the state each test has left after it has finished to identify where one test may be breaking the behavior of another. These are defined in RuleWatcher but could be eventually broken out into separate classes.

- IndexTracker -  Checks if the test has left items in the dist_reindex_journal queue, in particular if the queue is paused they may not be indexed for a while after the test has finished and indicates a problem in the test.
- configOverrideTracker - Tests if the test overrides any configuration with Config.setProperty,  if this is used and is not a unique property for the test then the test must reset it after or it may effect subsequent runs.  Note A test may set the property, and it was previously null which is fine as long as the code default is tshe same as the value set.
- transaction tracker - makes sure that the code did not manually start a transaction and not close it,  this will cause inconsistent behavior and could impact following tests.  Most of the time a transaction should have been automatically closed at the same place in the code it was opened in a finally block and @WrapInTransaction or LocalTransaction.wrap should be used to manage this. 

The output for these trackers can be seen in the test logs

General Fixes
- Proper waiting in tests for test content to be indexed on creation using the awaitability library to prevent timing issues
- ReindexThread.pause has been modified to trigger a transaction listener.  When an item was being added to the index queue table, the Reindexthread was being immediately notified to unpause, it would not see the new entries while the original transaction was still opened and would pause again after seeing no results requiring other code to trigger another unpause.  This requires consistent use of transaction listeners in transactions and a number of issues where transactions were not being closed or created manually have been resolved.  Quite a few upgrade tasks were not creating and closing transactions correctly when being run as a test individually, normally the wrapping runner was creating the transaction
- Initialize Integration tests in Runner. A lot of our tests use @DataProvider to set up individual test cases.  A lot of these Annotated Data provider methods use DotCMS APIs to set up the tests and require the Database to be initialized for this to work.  Rather than fix and rely on individual tests to initialized the IntegrationTestInitService we can have the Runner set this up before any of the DataProviders are called so all tests in the MainSuite will not fail.  The use of APIs in the DataProvider is an anti-pattern we should move away from longer term, the data for the test runs should be independent of the system itself, and it is better to use a static flag or a lazy loaded property for the APIs required to be called at runtime.
- Logging, we need to increase the log file size to be able to see the results.  We are currently using a RollingFile Appender, but we are unable to get hold of the subsequent files currently if the size rolls over,  the max file size that git can handle is 100MB

When Items are picked up to be indexed by the ReindexThread they are added to a BulkIndexer to send to Elasticsearch.  The items may not be delivered immediately.   Some test code is trying to poll whether an item has made it to the Elasticsearch Index or not by making a query  until the expected result is obtained.   This does not work well in all cases as the first time the same query is requested, the ESQueryCache caches the result any change will not be picked up until something else flushes the cache.  The cache can be manually cleared in the test or config property can be set to disable the cache.  

In progress

- Add Awaitability to replace Thread.sleep for test polling
- fixing testRemoveContentFromIndexMultilingualContent
- Fix test_content_type_reindex test handle duplicate entries
- make LocalTransaction consistent with commit listeners
- use commit listener for index unpause
- fix indexer pausing early
- fix session connection closing
- fix reset of DEFAULT_CONTENT_TO_DEFAULT_LANGUAGE
- handle exception and close transaction
- ensure startup tasks always run in transaction
- Add temp config override tracking
- fix testBundlerHandler_UnpublishHost_Success
- Turn off AUTO_ASSIGN_WORKFLOW for integration tests as it breaks after transaction fixes Test_Get_Bulk_Actions_For_Contentlet_Ids
- await index queue empty
- Initialize Integration Tests in Runner
- fix CaffineCache over logging
- create single log file for cicd
- update log file size to see all log
- fix SAVE_CONTENTLET_AS_JSON not being reset
- ServerAPIImplTest leaves extra enabled index servers after
- Fix leftover index entries
- Missing DELETE_SITE SystemEventType called from ESContentletAPIImpi.deleteContetlet
- duplicate key due to time resolution in test_save_and_delete_pushed_assets_on_two_env
- Fix test_that_live_and_working_content_makes_it_into_the_index

### PR DESCRIPTION
**Pull request by spbolton**
Add Awaitability to replace Thread.sleep for test polling fixing testRemoveContentFromIndexMultilingualContent Fix test_content_type_reindex test handle duplicate entries make LocalTransaction consistent with commit listeners use commit listener for index unpause
fix indexer pausing early
fix session connection closing
fix reset of DEFAULT_CONTENT_TO_DEFAULT_LANGUAGE
handle exception and close transaction
ensure startup tasks always run in transaction
Add temp config override tracking
fix testBundlerHandler_UnpublishHost_Success
Turn off AUTO_ASSIGN_WORKFLOW for integration tests as it breaks after transaction fixes Test_Get_Bulk_Actions_For_Contentlet_Ids await index queue empty
Initialize Integration Tests in Runner
fix CaffineCache over logging
create single log file for cicd
update log file size to see all log
fix SAVE_CONTENTLET_AS_JSON not being reset
ServerAPIImplTest leaves extra enabled index servers after Fix leftover index entries
Missing DELETE_SITE SystemEventType called from ESContentletAPIImpi.deleteContetlet duplicate key due to time resolution in test_save_and_delete_pushed_assets_on_two_env Fix test_that_live_and_working_content_makes_it_into_the_index

### CODE REVIEW

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImplTest.java [hunk 1]
```diff
@@ -645,271 +659,335 @@ public void testSearch () throws Exception {
...
+                user);
+
+        assert (liveSearchResults.get(0).getInode().equals(liveInode));
+
+        assert (workingSearchResults.get(0).getInode().equals(workingInode));
+        assert (!workingSearchResults.get(0).getInode().equals(liveInode));
+
+    }
+
+    public void waitForEmptyQueue() {
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImplTest.java:906]: why these methods `waitForEmptyQueue` , `pollSearchIndex` and `searchIndex` are public?

File: dotCMS/src/integration-test/java/com/dotcms/junit/RuleWatcher.java [hunk 2]
```diff
@@ -14,6 +34,100 @@ protected void starting(Description description) {
...
+                Logger.error(RuleWatcher.class,
+                        "Test " + description + " has open transaction after");
+                Object responseAction = Config.getStringProperty(
+                        TEST_TRANSACTION_TRACKER_AUTO_CLOSE_TYPE, TRANSACTION_AUTO_CLOSE_DEFAULT);
+                try {
+                    if (Config.getStringProperty(TEST_TRANSACTION_TRACKER_AUTO_CLOSE_TYPE,
+                            TRANSACTION_AUTO_CLOSE_DEFAULT).equals("commit")) {
+                        HibernateUtil.commitTransaction();
+                    } else {
+                        HibernateUtil.rollbackTransaction();
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/junit/RuleWatcher.java:115]: This else doesn't seem to be required here as you always rollback the transaction at the end

Comment (spbolton) [reply dotCMS/src/integration-test/java/com/dotcms/junit/RuleWatcher.java:115]: Call to second rollback should not be there, fixed

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImplTest.java (hunk 1 above)
Comment (spbolton) [reply dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImplTest.java]: Agree,  have updated

**Review by nollymar — APPROVED** [review]

**Review by fabrizzio-dotCMS — APPROVED** [review]

**Pull request merged**